tests/unit/pubsub_v1/publisher/test_flow_controller.py

```diff
@@ -0,0 +1,363 @@
+
+    _run_in_daemon(flow_controller, "add", [msg3], adding_3_done)
+    if adding_3_done.wait(timeout=0.1):
+        pytest.fail("Adding a message on overflow did not block.")
+
+    _run_in_daemon(flow_controller, "add", [msg4], adding_4_done)
+    if adding_4_done.wait(timeout=0.1):
+        pytest.fail("Adding a message on overflow did not block.")
+
+    # After releasing one message, there should be room for a new message, which
+    # should result in unblocking one of the waiting threads.
+    _run_in_daemon(flow_controller, "release", [msg1], releasing_1_done)
+    if not releasing_1_done.wait(timeout=0.1):
+        pytest.fail("Releasing a message blocked or errored.")
+
+    done_status = [
+        adding_2_done.wait(timeout=0.1),
+        adding_3_done.wait(timeout=0.1),
+        adding_4_done.wait(timeout=0.1),
+    ]
+
+    # In sum() we use the fact that True==1 and False==0, and that Event.wait()
+    # returns False only if it times out, i.e. its internal flag has not been set.
+    done_count = sum(done_status)
+    assert done_count == 1, "Exactly one thread should have been unblocked."
+
+    # Release another message and verify that yet another thread gets unblocked.
+    added_msg = [msg2, msg3, msg4][done_status.index(True)]
+    _run_in_daemon(flow_controller, "release", [added_msg], releasing_x_done)
+
+    if not releasing_x_done.wait(timeout=0.1):
+        pytest.fail("Releasing messages blocked or errored.")
+
+    released_count = sum(
+        (
+            adding_2_done.wait(timeout=0.1),
+            adding_3_done.wait(timeout=0.1),
+            adding_4_done.wait(timeout=0.1),
+        )
+    )
+    assert released_count == 2, "Exactly two threads should have been unblocked."
+
+
+def test_error_if_mesage_would_block_indefinitely():
+    settings = types.PublishFlowControl(
+        message_limit=0,  # simulate non-sane settings
+        byte_limit=1,
+        limit_exceeded_behavior=types.LimitExceededBehavior.BLOCK,
+    )
+    flow_controller = FlowController(settings)
+
+    msg = types.PubsubMessage(data=b"xyz")
+    adding_done = threading.Event()
+    error_event = threading.Event()
+
+    _run_in_daemon(flow_controller, "add", [msg], adding_done, error_event=error_event)
+
+    assert error_event.wait(timeout=0.1), "No error on adding too large a message."
+
+    # Now that we know that an error occurs, we can check its type directly
+    # without the fear of blocking indefinitely.
+    flow_controller = FlowController(settings)  # we want a fresh controller
+    with pytest.raises(exceptions.PermanentlyBlockedError) as error_info:
+        flow_controller.add(msg)
+
+    error_msg = str(error_info.value)
+    assert "messages: 1 / 0" in error_msg
+    assert "bytes: {} / 1".format(msg.ByteSize()) in error_msg
+
+
+def test_threads_posting_large_messages_do_not_starve():
+    settings = types.PublishFlowControl(
+        message_limit=100,  # simulate non-sane settings
```

Comment:
```suggestion
        message_limit=100,
```
Incorrect comment copied from another test case.